### PR TITLE
Customize our email backend

### PR DESCRIPTION
**Pull request by IegorSoboliev**

### CODE REVIEW

File: data_converter/settings.py [hunk 1]
```diff
@@ -163,6 +163,8 @@
 
 SITE_ID = 1
 
+EMAIL_BACKEND = 'email_backend.DataConverterEmailBackend'
```
Comment (roman-tiukh) [inline data_converter/settings.py:166]: you need to move this file to a directory, for example `data_converter`

File: email_backend.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+from sendgrid.helpers.mail import *
+
+
+class DataConverterEmailBackend(BaseEmailBackend):
+    def __init__(self, fail_silently=False, **kwargs):
+        super().__init__(fail_silently=fail_silently, **kwargs)
+        self.api_key = settings.SENDGRID_API_KEY
+
+    def send_messages(self, email_messages):
+        sg = sendgrid.SendGridAPIClient(api_key=settings.SENDGRID_API_KEY)
```
Comment (roman-tiukh) [inline email_backend.py:15]:
```suggestion
        sg = sendgrid.SendGridAPIClient(api_key=self.api_key)
```

File: email_backend.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        for email_message in email_messages:
+            from_email = Email(email_message.from_email)
+            to_email = To(email_message.to)
+            subject = 'Hi'
+            html_content = HtmlContent(email_message.body)
+            mail = Mail(from_email=from_email, to_emails=to_email, subject=subject, html_content=html_content)
+            sg.client.mail.send.post(request_body=mail.get())
+
+
+def send_template_mail(to_email, template, context, asynchronously=True):
```
Comment (roman-tiukh) [inline email_backend.py:25]: need `subject`, and `from_email`

Comment (roman-tiukh) [inline email_backend.py:25]:
```suggestion
def send_template_mail(to, subject, template, context, from_email=None, asynchronously=True):
```

File: email_backend.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    html = render_to_string(template, context)
+    e_message = EmailMessage(
+        body=html,
+        from_email=settings.DEFAULT_FROM_EMAIL,
+        to=[to_email],
+    )
```
Comment (roman-tiukh) [inline email_backend.py:31]:
```suggestion
    email = EmailMessage(
        body=html,
        from_email=settings.DEFAULT_FROM_EMAIL,
        to=to,  # to_email must be a list
    )
    email.content_subtype = 'html'
```

File: payment_system/templates/payment_system/new_subscription.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{% extends "users/emails/_base.html" %}
```
Comment (roman-tiukh) [inline payment_system/templates/payment_system/new_subscription.html:1]: move base template to general application `data_ocean/templates`

Comment (roman-tiukh) [inline payment_system/templates/payment_system/new_subscription.html:1]: move this file to subdirectory `emails`

payment_system/templates/payment_system/**emails**/new_subscription.html

File: data_converter/settings.py (hunk 1 above)
Comment (roman-tiukh) [reply data_converter/settings.py:166]: this is not a DataConverterEmailBackend, its SendGridEmailBackend

Comment (IegorSoboliev) [reply data_converter/settings.py:166]: No problem, renamed

File: email_backend.py (hunk 2 above)
Comment (IegorSoboliev) [reply email_backend.py:15]: done

File: email_backend.py (hunk 3 above)
Comment (IegorSoboliev) [reply email_backend.py:25]: added

File: email_backend.py (hunk 4 above)
Comment (IegorSoboliev) [reply email_backend.py:31]: added

File: payment_system/templates/payment_system/new_subscription.html (hunk 5 above)
Comment (IegorSoboliev) [reply payment_system/templates/payment_system/new_subscription.html:1]: done

File: data_converter/email_backend.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    def send_messages(self, email_messages):
+        sg = sendgrid.SendGridAPIClient(api_key=self.api_key)
+        for email_message in email_messages:
+            from_email = Email(email_message.from_email)
+            to_email = To(email_message.to)
+            subject = email_message.subject
+            html_content = HtmlContent(email_message.body)
+            mail = Mail(from_email=from_email, to_emails=to_email, subject=subject,
+                        html_content=html_content)
+            sg.client.mail.send.post(request_body=mail.get())
```
Comment (roman-tiukh) [inline data_converter/email_backend.py:24]: need to handle unsuccessful sending attempts

File: data_converter/email_backend.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def send_template_mail(to_email, subject, template, context, asynchronously=True,
+                       from_email=settings.DEFAULT_FROM_EMAIL):
```
Comment (roman-tiukh) [inline data_converter/email_backend.py:28]:
```suggestion
                       from_email=None):
```

File: data_converter/email_backend.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        subject=subject,
+        body=html,
+        from_email=from_email,
+        to=to_email,
+    )
+    e_message.content_subtype = 'html'
+    if not asynchronously:
+        e_message.send()
+    # then we have toe use Selery here
+    Thread(target=e_message.send()).start()
```
Comment (roman-tiukh) [inline data_converter/email_backend.py:40]:
```suggestion
    if not asynchronously:
        e_message.send()
    else:
        # then we have toe use Celery here
        Thread(target=e_message.send).start()
```

File: emails.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+from django.core.mail import EmailMessage, get_connection
+from django.conf import settings
+
```
Comment (roman-tiukh) [inline emails.py:3]: need to delete this file

File: sendgrid.env [hunk 10]
```diff
@@ -0,0 +1 @@
+
```
Comment (roman-tiukh) [inline sendgrid.env:1]: what is this?

File: requirements.txt [hunk 11]
```diff
@@ -43,8 +45,10 @@ lazy-object-proxy==1.4.3
 lxml==4.5.1
 MarkupSafe==1.1.1
 mccabe==0.6.1
+numpy==1.18.5
```
Comment (roman-tiukh) [inline requirements.txt:48]: numpy?

File: requirements.txt [hunk 12]
```diff
@@ -12,12 +12,14 @@ cffi==1.14.3
 chardet==3.0.4
 coreapi==2.3.3
 coreschema==0.0.4
+coverage==5.2.1
```
Comment (roman-tiukh) [inline requirements.txt:15]: coverage?

**Review by roman-tiukh — CHANGES_REQUESTED** [review]
`requirements.txt` must be more clear

File: requirements.txt (hunk 12 above)
Comment (IegorSoboliev) [reply requirements.txt:15]: this is for testing

Comment (roman-tiukh) [reply requirements.txt:15]: But not in this PR.

File: requirements.txt (hunk 11 above)
Comment (IegorSoboliev) [reply requirements.txt:48]: deleted

File: sendgrid.env (hunk 10 above)
Comment (IegorSoboliev) [reply sendgrid.env:1]: a recommendation from SendGrid

File: emails.py (hunk 9 above)
Comment (IegorSoboliev) [reply emails.py:3]: I am going to put all EmailMessages here

File: data_converter/email_backend.py (hunk 6 above)
Comment (IegorSoboliev) [reply data_converter/email_backend.py:24]: Is this good example from smtp.EmailBackend? `        try:
            self.connection.sendmail(from_email, recipients, message.as_bytes(linesep='\r\n'))
        except smtplib.SMTPException:
            if not self.fail_silently:
                raise
            return False
        return True
`

Comment (IegorSoboliev) [reply data_converter/email_backend.py:24]: And this is a SendGrid`s example `  import os
  from sendgrid import SendGridAPIClient
  from sendgrid.helpers.mail import (From, To, Subject, PlainTextContent, HtmlContent, Mail)
  from python_http_client import exceptions

  sendgrid_client = SendGridAPIClient(os.environ.get('SENDGRID_API_KEY'))
  from_email = From("help@twilio.com")
  to_email = To("ethomas@twilio.com")
  subject = Subject("Sending with Twilio SendGrid is Fun")
  plain_text_content = PlainTextContent("and easy to do anywhere, even with Python")
  html_content = HtmlContent("<strong>and easy to do anywhere, even with Python</strong>")
  message = Mail(from_email, to_email, subject, plain_text_content, html_content)
  try:
      response = sendgrid_client.send(message)
      print(response.status_code)
      print(response.body)
      print(response.headers)
  except exceptions.BadRequestsError as e:
      print(e.body)
      exit()`

File: payment_system/templates/payment_system/new_subscription.html (hunk 5 above)
Comment (IegorSoboliev) [reply payment_system/templates/payment_system/new_subscription.html]: done with new one

File: requirements.txt (hunk 12 above)
Comment (IegorSoboliev) [reply requirements.txt:15]: deleted

File: data_converter/email_backend.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        body=html,
+        from_email=from_email,
+        to=to,
+    )
+    e_message.content_subtype = 'html'
+    if not asynchronously:
+        e_message.send()
+    else:
+        # we have toe use Selery here
+        Thread(target=e_message.send()).start()
```
Comment (roman-tiukh) [inline data_converter/email_backend.py:53]: forgot to fix here after previous comments
```suggestion
        Thread(target=e_message.send).start()
```

File: data_converter/emails.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+EMAIL_CONFIRMATION = 'Підтвердження реєстрації у Data Ocean'
+
+
+def send_email_confirmation_message(to, user_name, confirm_link):
+    send_template_mail(to=[to, ],
+                       subject=EMAIL_CONFIRMATION,
+                       template='users/emails/email_confirmation.html',
+                       context={
+                           'user_name': user_name,
```
Comment (roman-tiukh) [inline data_converter/emails.py:12]: **Just for information**:
in this function, we can accept the user object as the argument to, `to = [to_user.email,]`
and we can pass the user object to the template, then call `{{user.get_full_name}}` in the template

File: data_converter/settings.py [hunk 15]
```diff
@@ -163,6 +163,8 @@
 
 SITE_ID = 1
 
+EMAIL_BACKEND = 'data_converter.email_backend.SendGridEmailBackend'
+
```
Comment (roman-tiukh) [inline data_converter/settings.py:167]: I think this option should be placed in settings_local.base.py

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: data_converter/settings.py (hunk 15 above)
Comment (IegorSoboliev) [reply data_converter/settings.py:167]: That is not a local setting

Comment (roman-tiukh) [reply data_converter/settings.py:167]: what if we need to install another EmailBackend for the test server? Also for development on a local machine, I want to install the console EmailBackend

File: data_converter/email_backend.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+            html_content = HtmlContent(email_message.body)
+            mail = Mail(from_email=from_email, to_emails=to_email, subject=subject,
+                        html_content=html_content)
+            try:
+                response = sg.client.mail.send.post(request_body=mail.get())
+                if response.status_code // 200 != 1:
+                    logger.warning(f'email "{subject}" was not sent to{email_message.to}. '
+                                   f'Status is {response.status_code}')
+            except HTTPError as error:
+                logger.error(f'email {subject} was not sent to{email_message.to} because of {error}')
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply data_converter/email_backend.py:36]: done

File: sendgrid.env (hunk 10 above)
Comment (IegorSoboliev) [reply sendgrid.env:1]: deleted

File: requirements.txt [hunk 17]
```diff
@@ -45,6 +46,7 @@ MarkupSafe==1.1.1
 mccabe==0.6.1
 oauthlib==3.1.0
 packaging==20.4
+pandas==1.0.4
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply requirements.txt:49]: deleted

File: data_ocean/templates/data_ocean/emails/_base.html [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    <img src="https://dataocean.us/static/img/logo.png" alt="DataOcean" width=106>
+    <hr style="margin-top: 15px;">
+    {% block content %}{% endblock %}
+    <hr>
+    <div class="bottom">
+        <a href=""><img src="linkedin.png" style="width: 20px;margin-right: 8px;display:inline;" width="20"
+                        height="20"/></a>
+        <a href=""><img src="facebook.png" style="width: 20px;margin-right: 8px;display:inline;" width="20"
+                        height="20"/></a>
+        <a href=""><img src="mail.png" style="width: 20px;display: inline;" width="20" height="20"/></a>
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply data_ocean/templates/data_ocean/emails/_base.html:24]: Olexandr is going to fix that later

File: data_converter/emails.py [hunk 19]
```diff
@@ -0,0 +1,15 @@
+from django.conf import settings
+from data_converter.email_backend import send_template_mail
+
+EMAIL_CONFIRMATION = 'Підтвердження реєстрації у Data Ocean'
+
+
+def send_email_confirmation_message(to, user_name, confirm_link):
+    send_template_mail(to=[to, ],
+                       subject=EMAIL_CONFIRMATION,
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply data_converter/emails.py:9]: deleted

File: data_converter/email_backend.py [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+        for email_message in email_messages:
+            from_email = Email(email_message.from_email)
+            to_email = To(email_message.to)
+            subject = email_message.subject
+            html_content = HtmlContent(email_message.body)
+            mail = Mail(from_email=from_email, to_emails=to_email, subject=subject,
+                        html_content=html_content)
+            try:
+                response = sg.client.mail.send.post(request_body=mail.get())
+                if response.status_code // 200 != 1:
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply data_converter/email_backend.py:32]: Thank you, changed

File: data_converter/email_backend.py [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+                    logger.warning(f'email "{subject}" was not sent to{email_message.to}. '
+                                   f'Status is {response.status_code}')
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply data_converter/email_backend.py:34]: Ok, changed

File: data_converter/email_backend.py (hunk 13 above)
Comment (IegorSoboliev) [reply data_converter/email_backend.py:53]: fixed

File: data_converter/emails.py (hunk 14 above)
Comment (IegorSoboliev) [reply data_converter/emails.py:12]: Let`s decide this when we will be defining places to call all send_..._message functions

File: data_converter/settings.py (hunk 15 above)
Comment (IegorSoboliev) [reply data_converter/settings.py:167]: ok, changed

**Review by roman-tiukh — APPROVED** [review]

File: data_converter/email_backend.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+class SendGridEmailBackend(BaseEmailBackend):
+    def __init__(self, fail_silently=False, **kwargs):
+        super().__init__(fail_silently=fail_silently, **kwargs)
+        self.api_key = settings.SENDGRID_API_KEY
+
+    def send_messages(self, email_messages):
+        sg = sendgrid.SendGridAPIClient(api_key=self.api_key)
+        for email_message in email_messages:
+            from_email = Email(email_message.from_email)
+            to_email = To(email_message.to)
```
[the remark this answers is not in the document]
Comment (roman-tiukh) [reply data_converter/email_backend.py:25]: ?

File: data_converter/email_backend.py [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+        subject=subject,
+        body=html,
+        from_email=from_email,
+        to=to,
+    )
+    e_message.content_subtype = 'html'
+    if not asynchronously:
+        e_message.send()
+    else:
+        # we have toe use Selery here
```
Comment (roman-tiukh) [inline data_converter/email_backend.py:52]:
```suggestion
        # we have toe use Celery here
```

**Review by roman-tiukh — APPROVED** [review]

File: data_converter/emails.py [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+                       subject='Новий рахунок',
+                       template='payment_system/emails/new_invoice.html',
+                       context={
+                           'user_name': user.first_name,
+                           'subscription_name': subscription_name,
+                           'start_date': start_date,
+                           'expiring_date': expiring_date,
+                           'invoice_link': invoice_link
+                       }
+                       )
```
Comment (roman-tiukh) [inline data_converter/emails.py:139]: **Better practice ->**

in template: `{{project_subscription.expiring_date}} {{project_subscription.start_date}}` 

```suggestion
def send_new_invoice_message(user, invoice):
    send_template_mail(
        to=[user.email],
        subject='Новий рахунок',
        template='payment_system/emails/new_invoice.html',
        context={
            'user': user,
            'project_subscription': invoice.project_subscription,
            'invoice': invoice,
        }
)
```

File: data_converter/emails.py [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+                           'member_first_name': member.first_name,
+                           'member_last_name': member.last_name,
+                           'project_name': project_name
+                       }
+                       )
+
+
+def send_membership_removed_message(user, owner, project_name):
+    send_template_mail(to=[user.email, ],
+                       subject='Вас видалили із проєкту',
```
Comment (roman-tiukh) [inline data_converter/emails.py:117]: @criticus @IegorSoboliev 
Maybe need to add like `[DataOcean]` prefix to all subjects?

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: data_converter/emails.py (hunk 25 above)
Comment (IegorSoboliev) [reply data_converter/emails.py:117]: Could you, please, explain the idea more?

File: data_converter/emails.py (hunk 24 above)
Comment (IegorSoboliev) [reply data_converter/emails.py:139]: changed except invoice_link

File: data_converter/emails.py [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+                       template='payment_system/emails/membership_removed.html',
+                       context={
+                           'user_name': user.first_name,
+                           'owner_first_name': owner.first_name,
+                           'owner_last_name': owner.last_name,
+                           'project_name': project_name
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply data_converter/emails.py:123]: used get_full_name function, but leave the assignment in the context for the simpler code

File: data_converter/emails.py [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+def send_registration_confirmed_message(user, default_project_link):
+    send_template_mail(to=[user.email, ],
+                       subject='Вітаємо у Data Ocean!',
+                       template='users/emails/registration_confirmed.html',
+                       context={
+                           'user_name': user.first_name,
+                           'default_project_link': default_project_link,
+                           'tariffs_link': TARIFFS_LINK
+                       }
+                       )
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply data_converter/emails.py:26]: changed

File: data_converter/email_backend.py (hunk 23 above)
Comment (IegorSoboliev) [reply data_converter/email_backend.py:52]: changed)

**Review by roman-tiukh — APPROVED** [review]

**Pull request merged**